lib/screens/v2/widget_instance/elevator_closures.ex

```diff
@@ -32,15 +25,111 @@ defmodule Screens.V2.WidgetInstance.ElevatorClosures do
           }
   end
 
+  defmodule Upcoming do
+    @moduledoc false
+
+    @enforce_keys ~w[period]a
+    defstruct @enforce_keys ++ ~w[summary]a
+
+    @type t :: %__MODULE__{period: {Date.t(), Date.t() | nil}, summary: String.t() | nil}
+
+    @nbsp "\u00A0"
+
+    @spec serialize(t(), DateTime.t()) :: map()
+    def serialize(%__MODULE__{period: {start_date, end_date}, summary: summary}, now) do
+      today = Util.service_date(now)
+      indefinite? = is_nil(end_date)
+      starts_tomorrow? = Date.diff(start_date, today) == 1
+      starts_this_week? = Date.beginning_of_week(start_date) == Date.beginning_of_week(today)
+      relative_start? = starts_tomorrow? or starts_this_week?
+      single_day? = start_date == end_date
+
+      same_month? =
+        not is_nil(end_date) and
+          {start_date.year, start_date.month} == {end_date.year, end_date.month}
+
+      {banner_title, details_titles} =
+        cond do
+          starts_tomorrow? ->
+            both_titles("Tomorrow")
+
+          starts_this_week? ->
+            both_titles("This #{day_name(start_date)}")
+
+          true ->
+            {
+              "#{day_name(start_date)}, #{short(start_date)}",
+              [
+                "#{day_name(start_date)}, #{long(start_date)}",
+                "#{day_name(start_date)}, #{short(start_date, _dot = true)}"
+              ]
+            }
+        end
+
+      {banner_trailer, details_trailer} =
+        cond do
+          indefinite? ->
+            both("until further notice")
+
+          single_day? ->
+            if relative_start?, do: {short(start_date), long(start_date)}, else: both(nil)
+
+          true ->
+            banner =
+              if same_month?,
+                do: "#{short(start_date)} – #{end_date.day}",
+                else: "#{short(start_date)} – #{short(end_date)}"
+
+            details =
+              if relative_start?,
+                do:
+                  if(same_month?,
+                    do: "#{long(start_date)} – #{end_date.day}",
+                    else: "#{long(start_date)} – #{long(end_date)}"
+                  ),
+                else: "through #{long(end_date)}"
+
+            {banner, details}
+        end
+
+      %{
+        banner: %{title: banner_title, trailer: banner_trailer},
+        details: %{summary: summary, titles: details_titles, trailer: details_trailer}
+      }
```

Comment:
Is there a reason you don't want to use a struct here? Since it's a nested map, that seems safer to me

Reply:
This is ultimately part of the return value of `serialize/1` on a widget, which is the last step before this data is encoded into JSON and sent to the client, so no further Elixir code would be using the value. In a way the "struct"/authoritative type definition for this data lives in the prop types of the client code.

This is mostly just observation of the conventions we already have in the codebase, so it's not to say we shouldn't change it (except to argue from consistency). I do feel vaguely as though using structs for this would give a _false_ notion of type-safety under the circumstances.

Reply:
Gotcha, since this is just getting sent off to the frontend, I see how there's less of a need for type safety from a struct